File: src/main/java/dev/slne/discord/spring/annotation/DiscordButton.java
package dev.slne.discord.spring.annotation;

import dev.slne.discord.annotation.DiscordEmoji;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import net.dv8tion.jda.api.interactions.components.buttons.ButtonStyle;
import org.springframework.stereotype.Component;

@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
@Component
public @interface DiscordButton {

  String id();

  String label();

  ButtonStyle style();

  DiscordEmoji emoji() default @DiscordEmoji;
}

File: src/main/kotlin/dev/slne/discord/annotation/DiscordButton.kt
package dev.slne.discord.annotation

import net.dv8tion.jda.api.components.buttons.Button
import net.dv8tion.jda.api.components.buttons.ButtonStyle
import org.springframework.stereotype.Component

@Target(AnnotationTarget.CLASS)
@Retention(AnnotationRetention.RUNTIME)
@Component
annotation class DiscordButton(
    val id: String,
    val label: String,
    val style: ButtonStyle,
    val emoji: DiscordEmoji = DiscordEmoji()
)

fun DiscordButton.toJdaButton() = Button.of(style, id, label, DiscordEmoji.Factory.create(emoji))
